src/details/ArborX_DetailsDistributedSearchTreeImpl.hpp

```diff
@@ -452,21 +452,24 @@ void DistributedSearchTreeImpl<DeviceType>::queryDispatch(
   ////////////////////////////////////////////////////////////////////////////
 }
 
-// FIXME: for some reason Kokkos::BinSort::sort() was not const.
-// If https://github.com/kokkos/kokkos/pull/1310 makes it into master in
-// Trilinos, we might want to pass bin_sort by const reference.
-template <typename BinSort>
-void applyPermutations(BinSort &)
+template <typename PermutationView>
+void applyPermutations(PermutationView const &)
 {
   // do nothing
 }
 
-template <typename BinSort, typename View, typename... OtherViews>
-void applyPermutations(BinSort &bin_sort, View view, OtherViews... other_views)
+template <typename PermutationView, typename View, typename... OtherViews>
+void applyPermutations(PermutationView const &permutation, View view,
+                       OtherViews... other_views)
 {
-  ARBORX_ASSERT(bin_sort.get_permute_vector().extent(0) == view.extent(0));
-  bin_sort.sort(view);
-  applyPermutations(bin_sort, other_views...);
+  ARBORX_ASSERT(permutation.extent(0) == view.extent(0));
+  View scratch(Kokkos::ViewAllocateWithoutInitializing("scratch"), view.size());
+  Kokkos::parallel_for(
+      "permute",
+      Kokkos::RangePolicy<typename View::execution_space>(0, view.size()),
+      KOKKOS_LAMBDA(int i) { scratch(i) = view(permutation(i)); });
+  Kokkos::deep_copy(view, scratch);
+  applyPermutations(permutation, other_views...);
```

Comment:
I wonder if there is any way you can apply permutation to multiple views simultaneously, maybe that could be a better performing.

Reply:
Unroll and demonstrate it is faster and then we will consider changing.

Reply:
No. I'm not arguing to do that here. I simply post a question.

Reply:
I don't know. After this pull request, we spend about 1% of the total runtime applying the permutation.